src/vmm/src/memory_dump.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Defines functionality for creating guest memory snapshots.
+
+use std::fmt::{Display, Formatter};
+use std::io::SeekFrom;
+
+use vm_memory::{
+    Bytes, GuestMemory, GuestMemoryError, GuestMemoryMmap, GuestMemoryRegion, MemoryRegionAddress,
+};
+
+use crate::DirtyBitmap;
+
+/// Defines the interface for dumping memory a file.
+pub trait DumpMemory {
+    fn dump<T: std::io::Write>(&self, writer: &mut T) -> std::result::Result<(), Error>;
+    fn dump_dirty<T: std::io::Write + std::io::Seek>(
+        &self,
+        writer: &mut T,
+        dirty_bitmap: &DirtyBitmap,
+    ) -> std::result::Result<(), Error>;
+}
+
+/// Errors associated with dumping guest memory to file.
+#[derive(Debug)]
+pub enum Error {
+    WriteMemory(GuestMemoryError),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        use self::Error::*;
+        match self {
+            WriteMemory(err) => write!(f, "Unable to dump memory: {:?}", err),
+        }
+    }
+}
+
+impl DumpMemory for GuestMemoryMmap {
+    fn dump<T: std::io::Write>(&self, writer: &mut T) -> std::result::Result<(), Error> {
+        self.with_regions_mut(|_, region| {
+            region
+                .write_to(MemoryRegionAddress(0), writer, region.len() as usize)
+                .map(|_| ())
+        })
+        .map_err(Error::WriteMemory)?;
+
+        Ok(())
+    }
+
+    fn dump_dirty<T: std::io::Write + std::io::Seek>(
+        &self,
+        writer: &mut T,
+        dirty_bitmap: &DirtyBitmap,
+    ) -> std::result::Result<(), Error> {
+        let mut writer_offset = 0;
+
+        self.with_regions_mut(|slot, region| {
+            let bitmap = dirty_bitmap.get(&slot).unwrap();
+
+            for (i, v) in bitmap.iter().enumerate() {
+                for j in 0..64 {
+                    let is_dirty_page = ((v >> j) & 1u64) != 0u64;
+                    if is_dirty_page {
+                        let page_offset = ((i * 64) + j) * 4096;
```

Comment:
It would be nice to define a const for PAGE_SIZE.

Reply:
Or get it from [`sysconf`](http://man7.org/linux/man-pages/man2/getpagesize.2.html).